README.md
# tiff-js

A Javascript interface for the libtiff image library, built on top of Node.js buffers using the "ref" type interface.

## Installation

Install with `npm`:

```bash
$ npm install tiff-js
```

## License

MIT License. See the [LICENSE](LICENSE) file.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.0.1] - 2019-07-13
	
### Added

- This CHANGELOG

### Fixed

- Build issues for macos and linux

[unreleased]: https://github.com/codenstein/tiff-js/compare/0.0.0...HEAD
[0.0.1]: https://github.com/codenstein/tiff-js/compare/0.0.0...0.0.1

test/tiffh.cc
#include <nan.h>
#include <string.h>
#include <limits.h>
#include "../tiff.h"

using namespace v8;

#undef max

// req'd since size_t are 64-bit values on 64-bit Windows operating systems and
// v8 does not support 64-bit numbers.
int SizeTToInt(size_t data)
{
    if (data > std::numeric_limits<int>::max())
    {
        fprintf(stderr, "%s", "Invalid cast. size_t exceeds the numeric limit of int.\n");
        abort();
    }
    return static_cast<int>(data);
}

#define STR(x) #x

#define ADD_TYPE(types, type)                                                                      \
    {                                                                                              \
        v8::Local<v8::Object> _type = Nan::New<v8::Object>();                                      \
        _type->Set(Nan::New("size").ToLocalChecked(), Nan::New<Number>(SizeTToInt(sizeof(type)))); \
        types->Set(Nan::New(STR(type)).ToLocalChecked(), _type);                                   \
    }

#define ADD_OFFSET(structs, type, member)                                                                                       \
    {                                                                                                                           \
        v8::Local<v8::Object> _type = Nan::To<v8::Object>(structs->Get(Nan::New(STR(type)).ToLocalChecked())).ToLocalChecked(); \
        v8::Local<v8::Object> offsets = Nan::To<v8::Object>(_type->Get(Nan::New("offsets").ToLocalChecked())).ToLocalChecked(); \
        offsets->Set(Nan::New(STR(member)).ToLocalChecked(), Nan::New<Number>(SizeTToInt(offsetof(type, member))));             \
    }

#define ADD_STRUCT(structs, type)                                                                    \
    {                                                                                                \
        v8::Local<v8::Object> _struct = Nan::New<v8::Object>();                                      \
        _struct->Set(Nan::New("size").ToLocalChecked(), Nan::New<Number>(SizeTToInt(sizeof(type)))); \
        _struct->Set(Nan::New("offsets").ToLocalChecked(), Nan::New<v8::Object>());                  \
        structs->Set(Nan::New(STR(type)).ToLocalChecked(), _struct);                                 \
    }

void InitTypedefs(v8::Local<v8::Object> exports)
{
    v8::Local<v8::Object> typedefs = Nan::New<v8::Object>();

    v8::Local<v8::Object> types = Nan::New<v8::Object>();
    ADD_TYPE(types, int8);
    ADD_TYPE(types, uint8);

    ADD_TYPE(types, int16);
    ADD_TYPE(types, uint16);

    ADD_TYPE(types, int32);
    ADD_TYPE(types, uint32);

    ADD_TYPE(types, int64);
    ADD_TYPE(types, uint64);

    ADD_TYPE(types, uint16_vap);
    typedefs->Set(Nan::New("types").ToLocalChecked(), types);

    v8::Local<v8::Object> structs = Nan::New<v8::Object>();

    /*
 * TIFF header.
 */
    /* Fixed point structure type. */
    ADD_STRUCT(structs, TIFFHeaderCommon);
    ADD_OFFSET(structs, TIFFHeaderCommon, tiff_magic);
    ADD_OFFSET(structs, TIFFHeaderCommon, tiff_version);

    /* Defines a frame rectangle in const ICAP_UNITS coordinates. */
    ADD_STRUCT(structs, TIFFHeaderClassic);
    ADD_OFFSET(structs, TIFFHeaderClassic, tiff_magic);
    ADD_OFFSET(structs, TIFFHeaderClassic, tiff_version);
    ADD_OFFSET(structs, TIFFHeaderClassic, tiff_diroff);

    /* Defines the parameters used for channel-specific transformation. */
    ADD_STRUCT(structs, TIFFHeaderBig);
    ADD_OFFSET(structs, TIFFHeaderBig, tiff_magic);
    ADD_OFFSET(structs, TIFFHeaderBig, tiff_version);
    ADD_OFFSET(structs, TIFFHeaderBig, tiff_offsetsize);
    ADD_OFFSET(structs, TIFFHeaderBig, tiff_unused);
    ADD_OFFSET(structs, TIFFHeaderBig, tiff_diroff);

    typedefs->Set(Nan::New("structs").ToLocalChecked(), structs);

    exports->Set(Nan::New("typedefs").ToLocalChecked(), typedefs);
}

NAN_MODULE_INIT(Init)
{
    InitTypedefs(target);
}

NODE_MODULE(tiffh, Init);

test/tiffh.js
var assert = require("assert");
var tiffh = require("bindings")({
  module_root: __dirname,
  bindings: "tiffh"
});
var tiffhJS = require("../lib/tiffh");

// need to create test cases that verify the sizes and offsets of structs
// from native tiff.h vs tiffh.js (ref-struct)

describe("Validate Type Definitions", function() {
  describe("Types", function() {
    describe("Sizes", function() {
      let types = Object.getOwnPropertyNames(tiffh.typedefs.types);
      let ignore = [];
      types.forEach(type => {
        if (ignore.includes(type)) {
          return;
        }
        it("#" + type, function() {
          assert.strictEqual(
            tiffhJS[type].size,
            tiffh.typedefs.types[type].size,
            "Expected sizes to be equal:"
          );
        });
      });
    });
  });
  describe("Structs", function() {
    describe("Sizes", function() {
      let structs = Object.getOwnPropertyNames(tiffh.typedefs.structs);
      structs.forEach(struct => {
        it("#" + struct, function() {
          assert.strictEqual(
            tiffhJS[struct].size,
            tiffh.typedefs.structs[struct].size,
            "Expected sizes to be equal:"
          );
        });
      });
    });
  });
});


tiffconf.h
/*
  Configuration defines for installed libtiff.
  This file maintained for backward compatibility. Do not use definitions
  from this file in your programs.
*/

#ifndef _TIFFCONF_
#define _TIFFCONF_

/* Define to 1 if the system has the type `int16'. */
/* #undef HAVE_INT16 */

/* Define to 1 if the system has the type `int32'. */
/* #undef HAVE_INT32 */

/* Define to 1 if the system has the type `int8'. */
/* #undef HAVE_INT8 */

/* The size of a `int', as computed by sizeof. */
#define SIZEOF_INT 4

/* Signed 8-bit type */
#define TIFF_INT8_T signed char

/* Unsigned 8-bit type */
#define TIFF_UINT8_T unsigned char

/* Signed 16-bit type */
#define TIFF_INT16_T signed short

/* Unsigned 16-bit type */
#define TIFF_UINT16_T unsigned short

/* Signed 32-bit type formatter */
#define TIFF_INT32_FORMAT "%d"

/* Signed 32-bit type */
#define TIFF_INT32_T signed int

/* Unsigned 32-bit type formatter */
#define TIFF_UINT32_FORMAT "%u"

/* Unsigned 32-bit type */
#define TIFF_UINT32_T unsigned int

/* Signed 64-bit type formatter */
#define TIFF_INT64_FORMAT "%I64d"

/* Signed 64-bit type */
#define TIFF_INT64_T int64_t // signed __int64

/* Unsigned 64-bit type formatter */
#define TIFF_UINT64_FORMAT "%I64u"

/* Unsigned 64-bit type */
#define TIFF_UINT64_T uint64_t // unsigned __int64

#if _WIN64
/*
  Windows 64-bit build
*/

/* Signed size type */
#  define TIFF_SSIZE_T TIFF_INT64_T

#else
/*
  Windows 32-bit build
*/

/* Signed size type */
#  define TIFF_SSIZE_T signed int

#endif

/* Compatibility stuff. */

/* Define as 0 or 1 according to the floating point format suported by the
   machine */
#define HAVE_IEEEFP 1

/* Set the native cpu bit order (FILLORDER_LSB2MSB or FILLORDER_MSB2LSB) */
#define HOST_FILLORDER FILLORDER_LSB2MSB

/* Native cpu byte order: 1 if big-endian (Motorola) or 0 if little-endian
   (Intel) */
#define HOST_BIGENDIAN 0

/* Support CCITT Group 3 & 4 algorithms */
#define CCITT_SUPPORT 1

/* Support JPEG compression (requires IJG JPEG library) */
/* #undef JPEG_SUPPORT */

/* Support JBIG compression (requires JBIG-KIT library) */
/* #undef JBIG_SUPPORT */

/* Support LogLuv high dynamic range encoding */
#define LOGLUV_SUPPORT 1

/* Support LZW algorithm */
#define LZW_SUPPORT 1

/* Support NeXT 2-bit RLE algorithm */
#define NEXT_SUPPORT 1

/* Support Old JPEG compresson (read contrib/ojpeg/README first! Compilation
   fails with unpatched IJG JPEG library) */
/* #undef OJPEG_SUPPORT */

/* Support Macintosh PackBits algorithm */
#define PACKBITS_SUPPORT 1

/* Support Pixar log-format algorithm (requires Zlib) */
/* #undef PIXARLOG_SUPPORT */

/* Support ThunderScan 4-bit RLE algorithm */
#define THUNDER_SUPPORT 1

/* Support Deflate compression */
/* #undef ZIP_SUPPORT */

/* Support strip chopping (whether or not to convert single-strip uncompressed
   images to mutiple strips of ~8Kb to reduce memory usage) */
#define STRIPCHOP_DEFAULT TIFF_STRIPCHOP

/* Enable SubIFD tag (330) support */
#define SUBIFD_SUPPORT 1

/* Treat extra sample as alpha (default enabled). The RGBA interface will
   treat a fourth sample with no EXTRASAMPLE_ value as being ASSOCALPHA. Many
   packages produce RGBA files but don't mark the alpha properly. */
#define DEFAULT_EXTRASAMPLE_AS_ALPHA 1

/* Pick up YCbCr subsampling info from the JPEG data stream to support files
   lacking the tag (default enabled). */
#define CHECK_JPEG_YCBCR_SUBSAMPLING 1

/* Support MS MDI magic number files as TIFF */
/* #undef MDI_SUPPORT */

/*
 * Feature support definitions.
 * XXX: These macros are obsoleted. Don't use them in your apps!
 * Macros stays here for backward compatibility and should be always defined.
 */
#define COLORIMETRY_SUPPORT
#define YCBCR_SUPPORT
#define CMYK_SUPPORT
#define ICC_SUPPORT
#define PHOTOSHOP_SUPPORT
#define IPTC_SUPPORT

#endif /* _TIFFCONF_ */
/*
 * Local Variables:
 * mode: c
 * c-basic-offset: 8
 * fill-column: 78
 * End:
 */


lib/tiffh.js
/*
 * Copyright (c) 1988-1997 Sam Leffler
 * Copyright (c) 1991-1997 Silicon Graphics, Inc.
 *
 * Permission to use, copy, modify, distribute, and sell this software and
 * its documentation for any purpose is hereby granted without fee, provided
 * that (i) the above copyright notices and this permission notice appear in
 * all copies of the software and related documentation, and (ii) the names of
 * Sam Leffler and Silicon Graphics may not be used in any advertising or
 * publicity relating to the software without the specific, prior written
 * permission of Sam Leffler and Silicon Graphics.
 *
 * THE SOFTWARE IS PROVIDED "AS-IS" AND WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS, IMPLIED OR OTHERWISE, INCLUDING WITHOUT LIMITATION, ANY
 * WARRANTY OF MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.
 *
 * IN NO EVENT SHALL SAM LEFFLER OR SILICON GRAPHICS BE LIABLE FOR
 * ANY SPECIAL, INCIDENTAL, INDIRECT OR CONSEQUENTIAL DAMAGES OF ANY KIND,
 * OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS,
 * WHETHER OR NOT ADVISED OF THE POSSIBILITY OF DAMAGE, AND ON ANY THEORY OF
 * LIABILITY, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE
 * OF THIS SOFTWARE.
 */

const tiffh = require("bindings")("tiffh");

const ffi = require("ffi");
const ref = require("ref");
ref.types.struct = require("ref-struct");

// tiffconf.h
const TIFF_INT8_T = ref.types.char;
const TIFF_UINT8_T = ref.types.uchar;

const TIFF_INT16_T = ref.types.short;
const TIFF_UINT16_T = ref.types.ushort;

const TIFF_INT32_T = ref.types.int32;
const TIFF_UINT32_T = ref.types.uint32;

const TIFF_INT64_T = ref.types.int64;
const TIFF_UINT64_T = ref.types.uint64;

/*
 * Intrinsic data types required by the file format:
 *
 * 8-bit quantities     int8/uint8
 * 16-bit quantities    int16/uint16
 * 32-bit quantities    int32/uint32
 * 64-bit quantities    int64/uint64
 * strings              unsigned char*
 */

const int8 = TIFF_INT8_T;
const uint8 = TIFF_UINT8_T;

const int16 = TIFF_INT16_T;
const uint16 = TIFF_UINT16_T;

const int32 = TIFF_INT32_T;
const uint32 = TIFF_UINT32_T;

const int64 = TIFF_INT64_T;
const uint64 = TIFF_UINT64_T;

/*
 * Some types as promoted in a variable argument list
 * We use uint16_vap rather then directly using int, because this way
 * we document the type we actually want to pass through, conceptually,
 * rather then confusing the issue by merely stating the type it gets
 * promoted to
 */

const uint16_vap = ref.types.int;

/*
 * TIFF header.
 */
const TIFFHeaderCommon = ref.types.struct({
  tiff_magic: uint16 /* magic number (defines byte order) */,
  tiff_version: uint16 /* TIFF version number */
});

const TIFFHeaderClassic = ref.types.struct({
  tiff_magic: uint16 /* magic number (defines byte order) */,
  tiff_version: uint16 /* TIFF version number */,
  tiff_diroff: uint32 /* byte offset to first directory */
});

const TIFFHeaderBig = ref.types.struct({
  tiff_magic: uint16 /* magic number (defines byte order) */,
  tiff_version: uint16 /* TIFF version number */,
  tiff_offsetsize: uint16 /* size of offsets, should be 8 */,
  tiff_unused: uint16 /* unused word, should be 0 */,
  tiff_diroff: uint64 /* byte offset to first directory */
});

/*
 * NB: In the comments below,
 *  - items marked with a + are obsoleted by revision 5.0,
 *  - items marked with a ! are introduced in revision 6.0.
 *  - items marked with a % are introduced post revision 6.0.
 *  - items marked with a $ are obsoleted by revision 6.0.
 *  - items marked with a & are introduced by Adobe DNG specification.
 */

/*
 * Tag data type information.
 *
 * Note: RATIONALs are the ratio of two 32-bit integer values.
 */
const TIFFDataType = {
  TIFF_NOTYPE: 0 /* placeholder */,
  TIFF_BYTE: 1 /* 8-bit unsigned integer */,
  TIFF_ASCII: 2 /* 8-bit bytes w/ last byte null */,
  TIFF_SHORT: 3 /* 16-bit unsigned integer */,
  TIFF_LONG: 4 /* 32-bit unsigned integer */,
  TIFF_RATIONAL: 5 /* 64-bit unsigned fraction */,
  TIFF_SBYTE: 6 /* !8-bit signed integer */,
  TIFF_UNDEFINED: 7 /* !8-bit untyped data */,
  TIFF_SSHORT: 8 /* !16-bit signed integer */,
  TIFF_SLONG: 9 /* !32-bit signed integer */,
  TIFF_SRATIONAL: 10 /* !64-bit signed fraction */,
  TIFF_FLOAT: 11 /* !32-bit IEEE floating point */,
  TIFF_DOUBLE: 12 /* !64-bit IEEE floating point */,
  TIFF_IFD: 13 /* %32-bit unsigned integer (offset) */,
  TIFF_LONG8: 16 /* BigTIFF 64-bit unsigned integer */,
  TIFF_SLONG8: 17 /* BigTIFF 64-bit signed integer */,
  TIFF_IFD8: 18 /* BigTIFF 64-bit unsigned integer (offset) */
};

/*
 * TIFF is defined as an incomplete type to hide the
 * library's internal data structures from clients.
 */
const TIFF = ref.types.struct();
const pTIFF = ref.refType(TIFF);

module.exports = {
  TIFF,
  pTIFF,

  TIFF_INT8_T,
  TIFF_UINT8_T,
  TIFF_INT16_T,
  TIFF_UINT16_T,
  TIFF_INT32_T,
  TIFF_UINT32_T,
  TIFF_INT64_T,
  TIFF_UINT64_T,

  int8,
  uint8,
  int16,
  uint16,
  int32,
  uint32,
  int64,
  uint64,

  uint16_vap,

  TIFFHeaderCommon,
  TIFFHeaderClassic,
  TIFFHeaderBig,

  TIFFDataType,

  // constants
  ...require("bindings")("tiffh").constants
};
